cmd/junitreport/pkg/builder/nested/test_suites_builder.go

```diff
@@ -0,0 +1,158 @@
+package nested
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/openshift/origin/cmd/junitreport/pkg/api"
+	"github.com/openshift/origin/cmd/junitreport/pkg/builder"
+	"github.com/openshift/origin/cmd/junitreport/pkg/errors"
+)
+
+// NewTestSuitesBuilder returns a new nested test suites builder. All test suites consumed
+// by this builder will be added to a multitree of suites rooted at the suites with the given names.
+func NewTestSuitesBuilder(rootSuiteNames []string) builder.TestSuitesBuilder {
+	rootSuites := []*api.TestSuite{}
+	for _, name := range rootSuiteNames {
+		rootSuites = append(rootSuites, &api.TestSuite{Name: name})
+	}
+
+	return &nestedTestSuitesBuilder{
+		testSuites: &api.TestSuites{
+			Suites: rootSuites,
+		},
+	}
+}
+
+const (
+	// TestSuiteNameDelimeter is the default delimeter for test suite names
+	TestSuiteNameDelimeter = "/"
+)
+
+// nestedTestSuitesBuilder is a test suites builder that nests suites under a root suite
+type nestedTestSuitesBuilder struct {
+	testSuites *api.TestSuites
+}
+
+// AddSuite adds a test suite to the test suites collection being built and updates the metrics
+// of any parent suites to subsume those of the new suite. If the parent of the test suite to be
+// added is found in the collection, the test suite is added as a child of that suite. Otherwise,
+// parent suites are created by successively removing one layer of package specificity until the
+// root name is found. For instance, if the suite named "root/package/subpackage/subsubpackage"
+// were to be added to an empty collection, the  suites named "root", "root/package", and "root/
+// package/subpackage" would be created and added first, then the suite could be added as a child
+// of the latter parent package. All test suites to be added are asssumed to be nested under one
+// of the root suites created by the constructor method and the attempted addition of a suite not
+// rooted in those suites will fail silently to allow for selective tree-building given a root.
+func (b *nestedTestSuitesBuilder) AddSuite(suite *api.TestSuite) error {
+	parent, err := b.findOrCreateParent(suite.Name)
+	if err != nil {
+		if errors.IsSuiteOutOfBoundsError(err) {
+			// if we were asked to add a new test suite that isn't a child of any current root,
+			// we ignore the request but do not throw an error
+			return nil
+		}
+		return err
+	}
+	parent.Children = append(parent.Children, suite)
+
+	b.updateMetrics(suite)
+
+	return nil
+}
+
+// findOrCreateParent will find or create the parent for the test suite with the given name
+func (b *nestedTestSuitesBuilder) findOrCreateParent(name string) (*api.TestSuite, error) {
+	if !b.isChildOfRoots(name) {
+		return nil, errors.NewSuiteOutOfBoundsError(name)
+	}
+	parentName, exists := b.getParentName(name)
```

Comment:
feels similar to https://golang.org/pkg/path/filepath/#Ext, just return "" for suites without parents